File: src/registry.rs
use std::{
    collections::BTreeMap,
    env, fs,
    io::{Cursor, Write},
    path::{Path, PathBuf},
    time::Duration,
};

#[cfg(unix)]
use std::os::unix::fs::PermissionsExt;

use anyhow::{Context, Result, anyhow, bail};
use reqwest::{Url, blocking::Client, header};
use semver::Version;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::{
    cache::Cache,
    resolved::registry_identity_parts,
    util::{atomic_write, flatten_single_directory, http_client},
};

const ACCEPT_JSON: &str = "application/vnd.swift.registry.v1+json";
const ACCEPT_ZIP: &str = "application/vnd.swift.registry.v1+zip";

#[derive(Debug, Clone)]
pub(crate) struct RegistryConfig {
    default_registry_url: Option<Url>,
    scoped_registry_urls: BTreeMap<String, Url>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct RegistryVersion {
    pub(crate) version: Version,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct RegistryVersionsCache {
    registry_url: String,
    identity: String,
    versions: Vec<RegistryVersion>,
}

#[derive(Debug, Deserialize)]
struct RegistriesFile {
    registries: Option<BTreeMap<String, RegistryEntry>>,
}

#[derive(Debug, Deserialize)]
struct RegistryEntry {
    url: String,
}

#[derive(Debug, Deserialize)]
struct ReleasesResponse {
    releases: BTreeMap<String, Release>,
}

#[derive(Debug, Deserialize)]
struct Release {
    problem: Option<serde_json::Value>,
}

#[derive(Debug, Deserialize)]
struct ReleaseInfo {
    resources: Vec<ReleaseResource>,
}

#[derive(Debug, Deserialize)]
struct ReleaseResource {
    name: String,
    #[serde(rename = "type")]
    content_type: String,
    checksum: String,
}

impl RegistryConfig {
    pub(crate) fn load(
        package_dir: &Path,
        config_path: Option<&Path>,
        default_registry_url: Option<&str>,
    ) -> Result<Self> {
        let mut config = Self {
            default_registry_url: None,
            scoped_registry_urls: BTreeMap::new(),
        };

        if let Some(global_path) = global_registries_path() {
            config.merge_file(&global_path)?;
        }

        config.merge_file(&package_dir.join(".swiftpm/configuration/registries.json"))?;

        if let Some(config_path) = config_path {
            config.merge_file(&registries_path_from_config_path(config_path))?;
        }

        if let Some(default_registry_url) = default_registry_url {
            config.default_registry_url = Some(parse_registry_url(default_registry_url)?);
        }

        Ok(config)
    }

    pub(crate) fn registry_url_for_identity(&self, identity: &str) -> Result<Url> {
        let (scope, _) = registry_identity_parts(identity)?;
        self.scoped_registry_urls
            .get(scope)
            .or(self.scoped_registry_urls.get(&scope.to_ascii_lowercase()))
            .or(self.default_registry_url.as_ref())
            .cloned()
            .ok_or_else(|| anyhow!("no registry configured for '{scope}' scope"))
    }

    fn merge_file(&mut self, path: &Path) -> Result<()> {
        if !path.exists() {
            return Ok(());
        }

        let file = fs::File::open(path)
            .with_context(|| format!("failed to open registry configuration {}", path.display()))?;
        let registries_file: RegistriesFile = serde_json::from_reader(file).with_context(|| {
            format!("failed to parse registry configuration {}", path.display())
        })?;
        let Some(registries) = registries_file.registries else {
            return Ok(());
        };

        for (scope, entry) in registries {
            let url = parse_registry_url(&entry.url)?;
            if scope == "[default]" {
                self.default_registry_url = Some(url);
            } else {
                self.scoped_registry_urls
                    .insert(scope.to_ascii_lowercase(), url);
            }
        }
        Ok(())
    }
}

pub(crate) fn registry_versions(
    identity: &str,
    registry_config: &RegistryConfig,
    cache: &Cache,
) -> Result<Vec<RegistryVersion>> {
    let registry_url = registry_config.registry_url_for_identity(identity)?;
    if let Some(versions) = read_cached_registry_versions(cache, registry_url.as_str(), identity)? {
        return Ok(versions);
    }

    let lock_key = format!("{registry_url}:{identity}");
    let _lock = cache.lock("registry-versions", &lock_key)?;
    if let Some(versions) = read_cached_registry_versions(cache, registry_url.as_str(), identity)? {
        return Ok(versions);
    }

    let versions = fetch_registry_versions(&registry_url, identity)?;
    write_cached_registry_versions(cache, registry_url.as_str(), identity, &versions)?;
    Ok(versions)
}

pub(crate) fn download_registry_archive(
    cache: &Cache,
    registry_config: &RegistryConfig,
    identity: &str,
    version: &str,
    destination: &Path,
) -> Result<()> {
    let registry_url = registry_config.registry_url_for_identity(identity)?;
    let archive_path = cache.registry_archive_path(identity, version);

    if !archive_path.exists() {
        let _lock = cache.lock("registry-archives", &archive_path.to_string_lossy())?;
        if !archive_path.exists() {
            let expected_checksum =
                fetch_source_archive_checksum(&registry_url, identity, version)?;
            let bytes = fetch_registry_archive(&registry_url, identity, version)?;
            verify_checksum(identity, version, &bytes, &expected_checksum)?;
            atomic_write(&archive_path, &bytes)?;
        }
    }

    let bytes = fs::read(&archive_path)?;
    extract_zip_archive(&bytes, destination)?;
    flatten_single_directory(destination)?;
    Ok(())
}

fn fetch_registry_versions(registry_url: &Url, identity: &str) -> Result<Vec<RegistryVersion>> {
    let response: ReleasesResponse = registry_client()
        .get(package_url(registry_url, identity, None)?)
        .header(header::ACCEPT, ACCEPT_JSON)
        .send()?
        .error_for_status()?
        .json()?;

    let mut versions = response
        .releases
        .into_iter()
        .filter(|(_, release)| release.problem.is_none())
        .filter_map(|(version, _)| Version::parse(&version).ok())
        .map(|version| RegistryVersion { version })
        .collect::<Vec<_>>();
    versions.sort_by(|left, right| left.version.cmp(&right.version));
    Ok(versions)
}

fn fetch_source_archive_checksum(
    registry_url: &Url,
    identity: &str,
    version: &str,
) -> Result<String> {
    let response: ReleaseInfo = registry_client()
        .get(package_url(registry_url, identity, Some(version))?)
        .header(header::ACCEPT, ACCEPT_JSON)
        .send()?
        .error_for_status()?
        .json()?;

    response
        .resources
        .into_iter()
        .find(|resource| {
            resource.name == "source-archive" && resource.content_type == "application/zip"
        })
        .map(|resource| resource.checksum)
        .ok_or_else(|| anyhow!("{identity} {version} does not declare a source archive checksum"))
}

fn fetch_registry_archive(registry_url: &Url, identity: &str, version: &str) -> Result<Vec<u8>> {
    let mut url = package_url(registry_url, identity, None)?;
    {
        let mut segments = url
            .path_segments_mut()
            .map_err(|_| anyhow!("registry URL cannot be a base: {registry_url}"))?;
        segments.push(&format!("{version}.zip"));
    }
    Ok(registry_client()
        .get(url)
        .header(header::ACCEPT, ACCEPT_ZIP)
        .send()?
        .error_for_status()?
        .bytes()?
        .to_vec())
}

fn verify_checksum(identity: &str, version: &str, bytes: &[u8], expected: &str) -> Result<()> {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    let actual = hex::encode(hasher.finalize());
    if !actual.eq_ignore_ascii_case(expected) {
        bail!("{identity} {version} checksum mismatch: expected {expected}, got {actual}");
    }
    Ok(())
}

fn extract_zip_archive(bytes: &[u8], destination: &Path) -> Result<()> {
    let reader = Cursor::new(bytes);
    let mut archive = zip::ZipArchive::new(reader)?;
    for index in 0..archive.len() {
        let mut file = archive.by_index(index)?;
        let Some(enclosed_name) = file.enclosed_name() else {
            bail!("registry archive contains an unsafe path: {}", file.name());
        };
        let path = destination.join(enclosed_name);
        if file.is_dir() {
            fs::create_dir_all(path)?;
        } else {
            if let Some(parent) = path.parent() {
                fs::create_dir_all(parent)?;
            }
            let mut output = fs::File::create(&path)?;
            std::io::copy(&mut file, &mut output)?;
            output.flush()?;
            #[cfg(unix)]
            if let Some(mode) = file.unix_mode() {
                fs::set_permissions(&path, fs::Permissions::from_mode(mode))?;
            }
        }
    }
    Ok(())
}

fn read_cached_registry_versions(
    cache: &Cache,
    registry_url: &str,
    identity: &str,
) -> Result<Option<Vec<RegistryVersion>>> {
    let path = cache.registry_versions_path(registry_url, identity);
    if !path.exists() {
        return Ok(None);
    }

    let metadata = fs::metadata(&path)?;
    let modified = metadata.modified()?;
    if modified.elapsed().unwrap_or(Duration::MAX) > Duration::from_secs(60 * 60) {
        return Ok(None);
    }

    let file = fs::File::open(&path)?;
    let cached: RegistryVersionsCache = serde_json::from_reader(file)
        .with_context(|| format!("failed to parse {}", path.display()))?;
    if cached.registry_url != registry_url || cached.identity != identity {
        return Ok(None);
    }
    Ok(Some(cached.versions))
}

fn write_cached_registry_versions(
    cache: &Cache,
    registry_url: &str,
    identity: &str,
    versions: &[RegistryVersion],
) -> Result<()> {
    let path = cache.registry_versions_path(registry_url, identity);
    let bytes = serde_json::to_vec_pretty(&RegistryVersionsCache {
        registry_url: registry_url.to_string(),
        identity: identity.to_string(),
        versions: versions.to_vec(),
    })?;
    atomic_write(&path, &bytes)?;
    Ok(())
}

fn package_url(registry_url: &Url, identity: &str, version: Option<&str>) -> Result<Url> {
    let (scope, name) = registry_identity_parts(identity)?;
    let mut url = registry_url.clone();
    {
        let mut segments = url
            .path_segments_mut()
            .map_err(|_| anyhow!("registry URL cannot be a base: {registry_url}"))?;
        segments.pop_if_empty();
        segments.push(scope);
        segments.push(name);
        if let Some(version) = version {
            segments.push(version);
        }
    }
    Ok(url)
}

fn registry_client() -> &'static Client {
    http_client()
}

fn parse_registry_url(url: &str) -> Result<Url> {
    let url = Url::parse(url)?;
    if url.scheme() != "https" {
        bail!("registry URL must use https: {url}");
    }
    Ok(url)
}

fn registries_path_from_config_path(config_path: &Path) -> PathBuf {
    if config_path.is_file() {
        config_path.to_path_buf()
    } else {
        config_path.join("registries.json")
    }
}

fn global_registries_path() -> Option<PathBuf> {
    env::var_os("HOME").map(|home| {
        PathBuf::from(home)
            .join(".swiftpm")
            .join("configuration")
            .join("registries.json")
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_default_and_scoped_registries_from_swiftpm_config() {
        let temp = tempfile::tempdir().unwrap();
        let config_dir = temp.path().join(".swiftpm/configuration");
        fs::create_dir_all(&config_dir).unwrap();
        fs::write(
            config_dir.join("registries.json"),
            r#"{
                "registries": {
                    "[default]": { "url": "https://registry.example.com/swift" },
                    "tuist": { "url": "https://registry.tuist.dev/api/registry/swift" }
                },
                "version": 1
            }"#,
        )
        .unwrap();

        let config = RegistryConfig::load(temp.path(), None, None).unwrap();

        assert_eq!(
            config
                .registry_url_for_identity("apple.swift-log")
                .unwrap()
                .as_str(),
            "https://registry.example.com/swift"
        );
        assert_eq!(
            config
                .registry_url_for_identity("tuist.FileSystem")
                .unwrap()
                .as_str(),
            "https://registry.tuist.dev/api/registry/swift"
        );
    }

    #[test]
    fn builds_package_urls_with_escaped_identity_parts() {
        let url = package_url(
            &Url::parse("https://registry.tuist.dev/api/registry/swift").unwrap(),
            "apple.swift-log",
            Some("1.12.0"),
        )
        .unwrap();

        assert_eq!(
            url.as_str(),
            "https://registry.tuist.dev/api/registry/swift/apple/swift-log/1.12.0"
        );
    }
}

File: Sources/swifterpm/Registry.swift
import Foundation

struct RegistryConfig: Sendable {
    private var defaultRegistryURL: URL?
    private var scopedRegistryURLs: [String: URL] = [:]

    static func load(packageDir: URL, configPath: URL?, defaultRegistryURL: String?) throws -> RegistryConfig {
        var config = RegistryConfig()
        if let global = globalRegistriesPath() {
            try config.mergeFile(global)
        }
        try config.mergeFile(packageDir.appendingPathComponent(".swiftpm/configuration/registries.json"))
        if let configPath {
            try config.mergeFile(registriesPath(fromConfigPath: configPath))
        }
        if let defaultRegistryURL {
            config.defaultRegistryURL = try parseRegistryURL(defaultRegistryURL)
        }
        return config
    }

    func registryURL(for identity: String) throws -> URL {
        let (scope, _) = try registryIdentityParts(identity)
        if let scoped = scopedRegistryURLs[scope] ?? scopedRegistryURLs[scope.lowercased()] {
            return scoped
        }
        if let defaultRegistryURL {
            return defaultRegistryURL
        }
        throw fail("no registry configured for '\(scope)' scope")
    }

    private mutating func mergeFile(_ path: URL) throws {
        guard FileManager.default.fileExists(atPath: path.path) else { return }
        guard let root = try JSONSerialization.jsonObject(with: Data(contentsOf: path)) as? [String: Any],
              let registries = root["registries"] as? [String: Any]
        else {
            return
        }
        for (scope, value) in registries {
            guard let entry = value as? [String: Any],
                  let urlString = entry["url"] as? String
            else {
                continue
            }
            let url = try parseRegistryURL(urlString)
            if scope == "[default]" {
                defaultRegistryURL = url
            } else {
                scopedRegistryURLs[scope.lowercased()] = url
            }
        }
    }
}

struct RegistryVersion: Codable, Sendable {
    let version: String

    var semver: SemVer? {
        try? SemVer(version)
    }
}

private struct RegistryVersionsCache: Codable {
    let registryURL: String
    let identity: String
    let versions: [RegistryVersion]
}

func registryVersions(identity: String, registryConfig: RegistryConfig, cache: Cache) async throws -> [RegistryVersion] {
    let registryURL = try registryConfig.registryURL(for: identity)
    if let cached = try readCachedRegistryVersions(cache: cache, registryURL: registryURL.absoluteString, identity: identity) {
        return cached
    }
    let lock = try cache.lock(namespace: "registry-versions", key: "\(registryURL.absoluteString):\(identity)")
    _ = lock
    if let cached = try readCachedRegistryVersions(cache: cache, registryURL: registryURL.absoluteString, identity: identity) {
        return cached
    }
    let versions = try await fetchRegistryVersions(registryURL: registryURL, identity: identity)
    try writeCachedRegistryVersions(cache: cache, registryURL: registryURL.absoluteString, identity: identity, versions: versions)
    return versions
}

func downloadRegistryArchive(
    cache: Cache,
    registryConfig: RegistryConfig,
    identity: String,
    version: String,
    destination: URL
) async throws {
    let registryURL = try registryConfig.registryURL(for: identity)
    let archivePath = cache.registryArchivePath(identity: identity, version: version)
    if !FileManager.default.fileExists(atPath: archivePath.path) {
        let lock = try cache.lock(namespace: "registry-archives", key: archivePath.path)
        _ = lock
        if !FileManager.default.fileExists(atPath: archivePath.path) {
            let expectedChecksum = try await fetchSourceArchiveChecksum(registryURL: registryURL, identity: identity, version: version)
            let data = try await fetchRegistryArchive(registryURL: registryURL, identity: identity, version: version)
            let actual = sha256Hex(data)
            guard actual.caseInsensitiveCompare(expectedChecksum) == .orderedSame else {
                throw fail("\(identity) \(version) checksum mismatch: expected \(expectedChecksum), got \(actual)")
            }
            try atomicWrite(data, to: archivePath)
        }
    }

    try runCommand("/usr/bin/unzip", ["-q", archivePath.path, "-d", destination.path])
    try flattenSingleDirectory(destination)
}

private func fetchRegistryVersions(registryURL: URL, identity: String) async throws -> [RegistryVersion] {
    struct ReleasesResponse: Decodable {
        struct Release: Decodable { let problem: String? }
        let releases: [String: Release]
    }
    let data = try await httpData(url: try packageURL(registryURL: registryURL, identity: identity), headers: ["Accept": "application/vnd.swift.registry.v1+json"])
    let response = try JSONDecoder().decode(ReleasesResponse.self, from: data)
    return response.releases.compactMap { version, release in
        guard release.problem == nil, let semver = try? SemVer(version) else { return nil }
        return RegistryVersion(version: semver.description)
    }.sorted { ($0.semver ?? SemVer(major: 0, minor: 0, patch: 0)) < ($1.semver ?? SemVer(major: 0, minor: 0, patch: 0)) }
}

private func fetchSourceArchiveChecksum(registryURL: URL, identity: String, version: String) async throws -> String {
    struct ReleaseInfo: Decodable {
        struct Resource: Decodable {
            let name: String
            let type: String
            let checksum: String
        }
        let resources: [Resource]
    }
    let data = try await httpData(url: try packageURL(registryURL: registryURL, identity: identity, version: version), headers: ["Accept": "application/vnd.swift.registry.v1+json"])
    let info = try JSONDecoder().decode(ReleaseInfo.self, from: data)
    guard let resource = info.resources.first(where: { $0.name == "source-archive" && $0.type == "application/zip" }) else {
        throw fail("\(identity) \(version) does not declare a source archive checksum")
    }
    return resource.checksum
}

private func fetchRegistryArchive(registryURL: URL, identity: String, version: String) async throws -> Data {
    let (scope, name) = try registryIdentityParts(identity)
    let url = registryURL.appendingPathComponents([scope, name, "\(version).zip"])
    return try await httpData(url: url, headers: ["Accept": "application/vnd.swift.registry.v1+zip"])
}

private func packageURL(registryURL: URL, identity: String, version: String? = nil) throws -> URL {
    let (scope, name) = try registryIdentityParts(identity)
    var components = [scope, name]
    if let version {
        components.append(version)
    }
    return registryURL.appendingPathComponents(components)
}

private func parseRegistryURL(_ url: String) throws -> URL {
    guard let parsed = URL(string: url), parsed.scheme == "https" else {
        throw fail("registry URL must use https: \(url)")
    }
    return parsed
}

private func registriesPath(fromConfigPath configPath: URL) -> URL {
    var isDirectory: ObjCBool = false
    if FileManager.default.fileExists(atPath: configPath.path, isDirectory: &isDirectory), !isDirectory.boolValue {
        return configPath
    }
    return configPath.appendingPathComponent("registries.json")
}

private func globalRegistriesPath() -> URL? {
    ProcessInfo.processInfo.environment["HOME"].map {
        URL(fileURLWithPath: $0).appendingPathComponent(".swiftpm/configuration/registries.json")
    }
}

private func readCachedRegistryVersions(cache: Cache, registryURL: String, identity: String) throws -> [RegistryVersion]? {
    let path = cache.registryVersionsPath(registryURL: registryURL, identity: identity)
    guard FileManager.default.fileExists(atPath: path.path) else { return nil }
    let attrs = try FileManager.default.attributesOfItem(atPath: path.path)
    if let modified = attrs[.modificationDate] as? Date,
       Date().timeIntervalSince(modified) > 60 * 60
    {
        return nil
    }
    let cached = try JSONDecoder().decode(RegistryVersionsCache.self, from: Data(contentsOf: path))
    guard cached.registryURL == registryURL, cached.identity == identity else { return nil }
    return cached.versions
}

private func writeCachedRegistryVersions(cache: Cache, registryURL: String, identity: String, versions: [RegistryVersion]) throws {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    let data = try encoder.encode(RegistryVersionsCache(registryURL: registryURL, identity: identity, versions: versions)) + Data("\n".utf8)
    try atomicWrite(data, to: cache.registryVersionsPath(registryURL: registryURL, identity: identity))
}
